File: runtimes/java-20.0/src/main/java/io/openruntimes/java/RuntimeRequest.java
package io.openruntimes.java;

import java.util.Map;

public class RuntimeRequest {
    private final String method;
    private final String scheme;
    private final String host;
    private final int port;
    private final String path;
    private final Map<String, String> query;
    private final String queryString;
    private final Map<String, String> headers;
    private final Object body;
    private final String bodyString;
    private final String url;

    public RuntimeRequest(
            String method,
            String scheme,
            String host,
            int port,
            String path,
            Map<String, String> query,
            String queryString,
            Map<String, String> headers,
            Object body,
            String bodyString,
            String url
    ) {
        this.method = method;
        this.scheme = scheme;
        this.host = host;
        this.port = port;
        this.path = path;
        this.query = query;
        this.queryString = queryString;
        this.headers = headers;
        this.body = body;
        this.bodyString = bodyString;
        this.url = url;
    }

    public String getMethod() {
        return method;
    }

    public String getScheme() {
        return scheme;
    }

    public String getHost() {
        return host;
    }

    public int getPort() {
        return port;
    }

    public String getPath() {
        return path;
    }

    public Map<String, String> getQuery() {
        return query;
    }

    public String getQueryString() {
        return queryString;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public Object getBody() {
        return body;
    }

    public String getBodyString() {
        return bodyString;
    }

    public String getUrl() {
        return url;
    }
}

File: runtimes/dotnet/versions/latest/src/RuntimeRequest.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DotNetRuntime
{
    public class RuntimeRequest
    {
        public string Method { get; private set; }
        public string Scheme { get; private set; }
        public string Host { get; private set; }
        public int Port { get; private set; }
        public string Path { get; private set; }
        public Dictionary<string, string> Query { get; private set; }
        public string QueryString { get; private set; }
        public string Url { get; private set; }
        public Dictionary<string, string> Headers { get; private set; }

        public object Body
        {
            get
            {
                var contentType = Headers.TryGetValue("content-type", out string? value)
                    ? (value ?? "").ToLower()
                    : "";

                if (contentType.StartsWith("application/json"))
                {
                    if (BodyBinary.Length == 0)
                    {
                        return new Dictionary<string, object?>();
                    }
                    else
                    {
                        return BodyJson;
                    }
                }

                return BodyText;
            }
        }
        public byte[] BodyBinary { get; private set; }
        public string BodyText =>
            System.Text.Encoding.UTF8.GetString(BodyBinary, 0, BodyBinary.Length);
        public string BodyRaw => BodyText;
        public Dictionary<string, object?> BodyJson =>
            JsonSerializer.Deserialize<Dictionary<string, object?>>(BodyText)
            ?? new Dictionary<string, object?>();

        public RuntimeRequest(
            string method,
            string scheme,
            string host,
            int port,
            string path,
            Dictionary<string, string> query,
            string queryString,
            string url,
            Dictionary<string, string> headers,
            byte[] bodyBinary
        )
        {
            BodyBinary = bodyBinary;
            Headers = headers;
            Method = method;
            Url = url;
            Host = host;
            Scheme = scheme;
            Path = path;
            QueryString = queryString;
            Query = query;
            Port = port;
        }
    }
}
